definition_validator/validator.cc

```diff
@@ -500,6 +500,23 @@ void validateSealed(core::Context ctx, const core::SymbolRef klass, const ast::C
     validateSealedAncestorHelper(ctx, singleton, classDef, klass, "extended");
 }
 
+void validateUselessRequiredAncestors(core::Context ctx, const core::SymbolRef sym) {
+    auto data = sym.data(ctx);
+
+    for (auto req : data->requiredAncestors(ctx)) {
+        if (data->derivesFrom(ctx, req.symbol)) {
+            if (auto e = ctx.state.beginError(req.loc, core::errors::Resolver::UselessRequiredAncestor)) {
+                e.setHeader("`{}` is already {} by `{}`", req.symbol.data(ctx)->show(ctx),
+                            req.symbol.data(ctx)->isClassOrModuleModule() ? "included" : "inherited", data->show(ctx));
+            }
+        }
+    }
+}
+
+void validateRequiredAncestors(core::Context ctx, const core::SymbolRef sym) {
+    validateUselessRequiredAncestors(ctx, sym);
+}
```

Comment:
It would be nice if this function could exit early if there were no required ancestors present on the symbol.

Reply:
We would also need to check the transitive ones to ensure all requirements are satisfied

Reply:
Could you write up an example for me?

Reply:
With all commits considered, `validateRequiredAncestors` does the following:

```cpp
void validateRequiredAncestors(core::Context ctx, const core::SymbolRef sym) {
    validateUselessRequiredAncestors(ctx, sym);
    validateUnsatisfiedRequiredAncestors(ctx, sym);
    validateUnsatisfiableClassAncestors(ctx, sym);
}
```

Where:

* `validateUselessRequiredAncestors` indeed needs only the local required ancestors
* `validateUnsatisfiedRequiredAncestors` needs the transitive ones (but early returns for modules and abstract classes)
* `validateUnsatisfiableClassAncestors` needs the required classes from the transitively required ancestors

Reply:
So for something like

```rb
module RA; end

module M
  extend T::Helpers
  requires_ancestor RA
end

class C1
  extend T::Helpers
  include M
end

class C2 < C1 # error: Should include RA
end
```

Even if `C2` has no locally required ancestors, we need to look in it's superclasses and mixins (`requiredAncestorsTransitive`) if one of them requires an ancestors to be able to display the error.